### PR TITLE
22186-eCompletion smart quotes is missing other common Smalltalk characters like | and <

### PR DESCRIPTION
**Pull request by macta**
Small refactor of smartCharactersMapping to put the smart characters in one place. Also adds missing | and < characterss as well as allowing the right hand matching character to be typed on a selection.

### CODE REVIEW

**Comment by guillep** [comment]
Hi @macta I understand how that would work for pragmas and temporaries. But what about expressions like

```
1 < 2
```

or

```
true | false
```

?

**Comment by macta** [comment]
@guillep you are correct - my previous code only applied the <> and  || when there was a selection - in refactoring away the duplication for strings for these characters (as they are all over the place), I've lost that distinction. I'll need a rethink

**Comment by macta** [comment]
@guillep this further PR properly handles < and | but more importantly adds tests for how it works now in an attempt to document what was already there. So I think this does help clean it up a bit and remove some duplication (even for what is a trivial change)

File: src/NECompletion-Tests/NECControllerTest.class.st [hunk 1]
```diff
@@ -5,27 +5,205 @@ Class {
...
 
+{ #category : #'tests-keyboard' }
+NECControllerTest >> allSelfClosingCharacters [
+	^ {'{}' . '[]' . '()' . '<>' . '||' . '""' . ''''''}
+]
+
+{ #category : #'tests-keyboard' }
+NECControllerTest >> allSmartCharacters [
+	^ self allSelfClosingCharacters , { '||'. '<>' }
+]
```
Comment (guillep) [inline src/NECompletion-Tests/NECControllerTest.class.st:22]: Isn't this a duplication? I mean, the '||' and '<>' are twice there...

**Review by guillep — CHANGES_REQUESTED** [review]
Otherwise than my comment above, I think a cleanup on eCompletion is good...
I don't quite understand the implications of this change and how it may break typing (if it does).

I propose to solve this duplication and then integrate it and listen to people's comments. We can always revert if something goes wrong :)

File: src/NECompletion-Tests/NECControllerTest.class.st (hunk 1 above)
Comment (macta) [reply src/NECompletion-Tests/NECControllerTest.class.st:22]: I'm not sure what you mean - this is in the test so we can run through the expected combinations of matching characters?

I could of course expose the list of characters as a collection in NecController itself (its private at the moment - hence why the test has another list). It did cross my mind - what do you think?

Also - these characters are in a dictionary in NecController - as there was already a bit of jiggery to match closing characters - it did occur to me to just have a list of "SmartCharacter objects", so each smart characters knows its start and end value and whether it should complete. I wasn't sure if this was over egging it though?

At least with the tests I wrote, we could explore that option. And happy to do that if you think its better.

There is still some duplication in TextEditor, but I don't think that is actually used as Calypso is using RubTextEditor.

Comment (macta) [reply src/NECompletion-Tests/NECControllerTest.class.st:22]: To be clear - this change should be invisible to everyone. You will only notice a difference if you highlight some text and press | or < (and any of the existing closing characters ],},)

Comment (macta) [reply src/NECompletion-Tests/NECControllerTest.class.st:22]: What do i need to do to get this PR resolved?

Comment (pavel-krivanek) [reply src/NECompletion-Tests/NECControllerTest.class.st:22]: I restarted the tests check. It seems to work fine.

**Review by pavel-krivanek — APPROVED** [review]

**Review by MarcusDenker — APPROVED** [review]
I tested it, works!

**Pull request merged**